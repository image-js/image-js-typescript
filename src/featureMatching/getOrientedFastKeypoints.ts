import { Image } from '../Image';

import {
  FastKeypoint,
  getFastKeypoints,
  GetFastKeypointsOptions,
} from './getFastKeypoints';

export interface GetOrientedFastKeypointsOptions
  extends GetFastKeypointsOptions {}

export interface OrientedFastKeypoint extends FastKeypoint {
  /**
   * Anti-clockwise angle of the keypoint in degrees with regard to a horizontal line.
   */
  angle: number;
}

/**
 * Find the oriented FAST features in a GREY image.
 * How to add orientation to FAST is described in: http://www.gwylab.com/download/ORB_2012.pdf
 * Basically, the intensity centroid of the window around the corner is computed and the
 * orientation is given by the vector from the center to the intensity centroid.
 *
 * @param image - The image to process.
 * @param options - Get oriented FAST keypoints options.
 * @returns The oriented FAST keypoints.
 */
export function getOrientedFastKeypoints(
  image: Image,
  options: GetOrientedFastKeypointsOptions = {},
): OrientedFastKeypoint[] {
  const { windowSize = 7 } = options;

  const fastKeypoints = getFastKeypoints(image, options);
}
