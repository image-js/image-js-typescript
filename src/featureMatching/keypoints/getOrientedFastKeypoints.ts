import { Image } from '../../Image';
import { getAngle } from '../../maskAnalysis/utils/getAngle';
import { toDegrees } from '../../utils/geometry/angles';
import { getRadius } from '../../utils/getRadius';
import { checkBorderDistance } from '../utils/checkBorderDistance';

import {
  FastKeypoint,
  getFastKeypoints,
  GetFastKeypointsOptions,
} from './getFastKeypoints';
import { getPatchIntensityCentroid } from './getPatchIntensityCentroid';

export interface GetOrientedFastKeypointsOptions
  extends GetFastKeypointsOptions {
  /**
   * Window size for the intensity centroid computation.
   *
   * @default 7
   */
  windowSize?: number;
}

export interface OrientedFastKeypoint extends FastKeypoint {
  /**
   * Clockwise angle of the keypoint in degrees with regard to a horizontal line.
   */
  angle: number;
}

/**
 * Find the oriented FAST features in a GREY image.
 * How to add orientation to FAST is described in: http://www.gwylab.com/download/ORB_2012.pdf
 * Basically, the intensity centroid of the window around the corner is computed and the
 * orientation is given by the vector from the center to the intensity centroid.
 *
 * @param image - The image to process.
 * @param options - Get oriented FAST keypoints options.
 * @returns The oriented FAST keypoints.
 */
export function getOrientedFastKeypoints(
  image: Image,
  options: GetOrientedFastKeypointsOptions = {},
): OrientedFastKeypoint[] {
  const { windowSize = 7 } = options;

  const fastKeypoints = getFastKeypoints(image, options);

  const radius = getRadius(windowSize);

  // handle edge cases: remove keypoints too close to border
  let keypoints: FastKeypoint[] = [];
  for (let keypoint of fastKeypoints) {
    if (checkBorderDistance(image, keypoint.origin, radius)) {
      keypoints.push(keypoint);
    }
  }

  let orientedFastKeypoints: OrientedFastKeypoint[] = [];
  for (let keypoint of keypoints) {
    const centroid = getPatchIntensityCentroid(image, {
      origin: keypoint.origin,
      radius,
    })[0];
    const angle = toDegrees(getAngle({ column: 0, row: 0 }, centroid));
    orientedFastKeypoints.push({ ...keypoint, angle });
  }
  return orientedFastKeypoints;
}
