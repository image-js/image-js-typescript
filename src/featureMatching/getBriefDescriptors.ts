import { Image, ImageColorModel, ImageCoordinates } from '../Image';
import { GaussianBlurOptions } from '../filters';
import checkProcessable from '../utils/checkProcessable';
import { InterpolationType } from '../utils/interpolatePixel';

import { OrientedFastKeypoint } from './getOrientedFastKeypoints';
import { checkBorderDistance } from './utils/checkBorderDistance';
import { compareIntensity } from './utils/compareIntensity';
import { extractSquareImage } from './utils/extractSquareImage';
import {
  getGaussianPoints,
  GetGaussianPointsOptions,
} from './utils/getGaussianPoints';

export interface GetBriefDescriptorsOptions {
  /**
   * Options to smooth the image patch before comparing pairs of points.
   */
  smoothingOptions?: GaussianBlurOptions;
  /**
   * Options to modify the gaussian distribution used to generate the points to compare.
   */
  pointsDistributionOptions?: Omit<GetGaussianPointsOptions, 'nbPoints'>;
  /**
   * Size of the patch around the keypoint used to compute the descriptor.
   *
   * @default 31
   */
  patchSize?: number;
  /**
   * Number of bits of the final descriptor.
   *
   * @default 256
   */
  descriptorLength?: 128 | 256 | 512;
}

export type BriefDescriptor = Uint8Array;

/**
 * Generate the rBRIEF descriptors for the desired keypoints of an image.
 * The rBRIEF descriptors are presented in these articles:
 * - ORB article: DOI: 10.1109/ICCV.2011.6126544
 * - rBRIEF article: DOI: 10.1007/978-3-642-15561-1_56
 *
 * @param image - Source image of the keypoints.
 * @param keypoints - Keypoints for which the descriptors are wanted.
 * @param options - Get rotated BRIEF descriptors options
 * @returns The descriptors for the given keypoints.
 */
export function getBriefDescriptors(
  image: Image,
  keypoints: OrientedFastKeypoint[],
  options: GetBriefDescriptorsOptions = {},
): BriefDescriptor[] {
  const {
    patchSize = 31,
    descriptorLength = 256,
    smoothingOptions = { sigma: Math.sqrt(2), size: 9 },
    pointsDistributionOptions,
  } = options;

  checkProcessable(image, 'getBriefDescriptors', {
    alpha: false,
    colorModel: ImageColorModel.GREY,
  });

  if (!(patchSize % 2)) {
    throw new Error('getBriefDescriptors: patchSize should be an odd integer');
  }

  const gaussianPoints = getGaussianPoints(patchSize, patchSize, {
    ...pointsDistributionOptions,
    nbPoints: descriptorLength * 2,
  });

  const smoothed = image.gaussianBlur(smoothingOptions);

  const descriptors: Uint8Array[] = [];

  for (let keypoint of keypoints) {
    // crop smallest square surrounding the tilted patch of the keypoint
    // we have to handle the fact that this square can have even dimensions
    const rawWidth = Math.floor(
      patchSize *
        (Math.abs(Math.cos(keypoint.angle)) +
          Math.abs(Math.sin(keypoint.angle))),
    );

    const cropWidth = rawWidth % 2 ? rawWidth : rawWidth - 1;

    // ignore keypoints that are too close to the border of the image
    let borderDistance = (cropWidth - 1) / 2;
    if (!checkBorderDistance(smoothed, keypoint.origin, borderDistance)) {
      continue;
    }

    const cropped = extractSquareImage(smoothed, keypoint.origin, cropWidth);

    const rotateCenter = cropped.getCoordinates(ImageCoordinates.CENTER);
    const rotated = cropped.rotate(keypoint.angle, {
      center: rotateCenter,
      interpolationType: InterpolationType.NEAREST,
    });

    const cropOrigin = rotated.getCoordinates(ImageCoordinates.CENTER);
    const patch = extractSquareImage(rotated, cropOrigin, patchSize);

    const descriptor = new Uint8Array(descriptorLength);
    for (let i = 0; i < descriptorLength; i++) {
      const p1 = gaussianPoints[i];
      const p2 = gaussianPoints[i + descriptorLength];
      descriptor[i] = Number(compareIntensity(patch, p1, p2));
    }
    descriptors.push(descriptor);
  }

  return descriptors;
}
