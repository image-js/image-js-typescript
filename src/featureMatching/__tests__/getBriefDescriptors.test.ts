import { ImageColorModel } from '../../Image';
import { getBriefDescriptors } from '../getBriefDescriptors';
import { getOrientedFastKeypoints } from '../getOrientedFastKeypoints';

test('alphabet image, maxNbFeatures = 10', () => {
  const image = testUtils.load('various/alphabet.jpg');
  const grey = image.convertColor(ImageColorModel.GREY);

  const keypoints = getOrientedFastKeypoints(grey, { maxNbFeatures: 10 });
  expect(keypoints).toHaveLength(10);

  const keypointsCoordinates = keypoints.map((kpt) => kpt.origin);

  let imageWithKeypoints = image.clone();
  for (let i = 0; i < 10; i++) {
    imageWithKeypoints.drawCircle(keypointsCoordinates[i], 5, {
      color: [255, 0, 0, 255],
      out: imageWithKeypoints,
    });
  }

  expect(imageWithKeypoints).toMatchImageSnapshot();

  const descriptors = getBriefDescriptors(grey, keypoints);

  expect(descriptors).toMatchSnapshot();
});

test.only('should work with small patch size', () => {
  const image = testUtils.load('various/alphabet.jpg');
  const grey = image.convertColor(ImageColorModel.GREY);

  const keypoints = getOrientedFastKeypoints(grey, { maxNbFeatures: 10 });
  expect(keypoints).toHaveLength(10);

  const keypoint = keypoints.slice(0, 1);
  console.log(keypoint);

  const keypointCoordinates = keypoint.map((kpt) => kpt.origin);

  let imageWithKeypoints = image.clone();

  imageWithKeypoints.drawCircle(keypointCoordinates[0], 5, {
    color: [255, 0, 0, 255],
    out: imageWithKeypoints,
  });

  expect(imageWithKeypoints).toMatchImageSnapshot();

  const descriptor = getBriefDescriptors(grey, keypoint, { patchSize: 5 });

  console.log(descriptor);
  expect(descriptor).toMatchSnapshot();
});
