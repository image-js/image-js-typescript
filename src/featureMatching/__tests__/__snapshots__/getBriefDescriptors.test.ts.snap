// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should work with small patch size 2`] = `
[
  Uint8Array [
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    1,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    1,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    1,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    1,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    1,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    0,
    1,
    0,
    1,
  ],
]
`;
